### DISCUSSION TITLE
Unauthorized to create provider

### DISCUSSION THREAD

**Discussion by txrubel** [fonoster/yaps#242] [Q&A] [answered]
Hello,

getting Unauthorized error while trying create provider, Here I created new project and set it as default. And then I think CLI created a api secret for this project as following,
![image](https://user-images.githubusercontent.com/38957915/142673554-f2efd579-1313-4b0f-b5f3-fbfd2f648050.png)

and the error found in the CLI,
![image](https://user-images.githubusercontent.com/38957915/142674351-62199997-2a25-482e-992f-a9877a9aabe5.png)

and, from the debug log of fonoster_apiserver_1,
> time="2021-11-19T18:09:21Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"POST\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/fonoster.providers.v1beta1.Providers/C\
reateProvider\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/2.0\",\"ProtoMajor\":2,\"ProtoMinor\":0,\"Header\":{\"Accept-Encoding\":[\"identity\"],\"Access_key_id\":[\"PJ6197d53d0b118e\
0700000001\"],\"Access_key_secret\":[\"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJmb25vc3RlciIsInJvbGUiOiJQUk9KRUNUIiwiYWNjZXNzS2V5SWQiOiJQSjYxOTdkNTNkMGIxMThlMDcwMDAwMDAwMSIsImlhdCI6MTYzNzM0MDQ3NywiZXhwIjoxNjY4ODk4MDc3fQ.lM0Vy8N-31AsVloDhyFhFB9Nxr\
qJpfAVmCuL-_sGhuQ\"],\"Content-Type\":[\"application/grpc\"],\"Grpc-Accept-Encoding\":[\"identity,deflate,gzip\"],\"Te\":[\"trailers\"],\"User-Agent\":[\"grpc-node-js/1.4.4\"],\"X-Forwarded-Host\":[\"172.24.0.5:50051\"],\"X-Forwarded-Port\":[\"50051\"],\
\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"f8363bf8c487\"],\"X-Real-Ip\":[\"172.24.0.1\"]},\"ContentLength\":-1,\"TransferEncoding\":null,\"Host\":\"172.24.0.5:50051\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":nu\
ll,\"RemoteAddr\":\"172.24.0.1:43830\",\"RequestURI\":\"/fonoster.providers.v1beta1.Providers/CreateProvider\",\"TLS\":null}"

Its look like request made with correct secret.

From log in the fonoster_providers_1 container,
![image](https://user-images.githubusercontent.com/38957915/142675028-17f0334d-885c-416e-9100-dc8d3406e6b9.png)

Now my question, is there any place we need to add this secret before creating the provider?

Thanks,

**Comment by psanders** [comment, marked as answer]
Can you share the content of your `config/rbac.json`? Also, if you decode accessKeySecret at [ jwt.io]( jwt.io) what is the payload? 

-- Update --

We don't know what caused but might be related to [issue 245](https://github.com/fonoster/fonoster/issues/245). If you encounter this problem please send us your logs. The solution, in this case, was to clean the Docker engine and run the process again:

https://github.com/fonoster/fonoster/discussions/242#discussioncomment-1694219

**Reply by txrubel** [reply]
Here attached rbac.json file.
[rbac.json.txt](https://github.com/fonoster/fonoster/files/7572947/rbac.json.txt)

USER accessKeySecret payload:
> {
  "iss": "fonoster",
  "role": "USER",
  "accessKeyId": "fonoster",
  "iat": 1637301011,
  "exp": 1668858611
}

PROJECT accessKeySecret payload:
> {
  "iss": "fonoster",
  "role": "PROJECT",
  "accessKeyId": "PJ6197d53d0b118e0700000001",
  "iat": 1637340477,
  "exp": 1668898077
}

**Reply by psanders** [reply]
This all looks correct. What happens if you restart the service? It feels as if something went out of sync.

**Reply by txrubel** [reply]
Yeah I have reload all services, but no luck...
Found some error in the fonoster_fs_1 container, are those might be an issue?

![image](https://user-images.githubusercontent.com/38957915/142695277-3bcacb21-a14e-4825-9982-2e2f3da7ea44.png)

All following containers are running here,
![image](https://user-images.githubusercontent.com/38957915/142697556-96854eda-8ad7-4527-a76a-2c459bde3428.png)

**Reply by txrubel** [reply]
Now getting following error in providers container,
![image](https://user-images.githubusercontent.com/38957915/142711196-2721d6cf-7b9c-4085-adbd-e45310ddc69e.png)

**Reply by txrubel** [reply]
Maybe this "Bad token" is related to following code of auth container,

>  /**
   * Returns the decoded payload if the signature is valid even if it is expired
   */
  async decode(token: string, privateKey: string): Promise<JwtPayload> {
    try {
      // @ts-ignore
      return (await promisify(verify)(token, privateKey, {
        ignoreExpiration: false
      })) as JwtPayload;
    } catch (e) {
      logger.log("error", "@fonoster/auth [Bad token] " + token);
      throw new Error(e);
    }
  }

this file is read only cant do any change on it. I just wanted to see which token it reported wrong.

**Reply by txrubel** [reply]
one more thing to share, trying to access routr rest api using SIPPROXY_API_USERNAME and SIPPROXY_API_SECRET, it also unauthorized,
> curl -k -u "admin:bdfab48a4ce7d8d5f3dcb5c30dd63f09" https://172.24.0.3:4567/api/v1beta1/token
{"status": "401", "message":"Unauthorized"}

UPDATE: One interesting thing to share, 
While we do this rest-api request then found following response from AUTH container,
![image](https://user-images.githubusercontent.com/38957915/142737024-48fe6ac5-cad7-4f15-8064-e0b11268f286.png)
if we use that secret at the curl then get success response,
> curl -k -u "admin:ad8e42cde8acd7154929adee06db4ae7" https://172.24.0.3:4567/api/v1beta1/token
{"status":200,"message":"Successful request","data":"eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJhZG1pbiJ9.zobXvNV14hyVsYQg_LtlfBts5DxhW1DoIuwKyHYc2kb95xhfXE3YHEsl_hNoeMH2oVqft8tRShwPV8orhFd3Cg"}

**Reply by psanders** [reply]
Humm, this must be the root of the problem! I will test the installation process to see if we have a regression.

**Reply by psanders** [reply]
@txrubel Based on your description of the problem, I believe your installation process was interrupted. Please take a look at [issue 245](https://github.com/fonoster/fonoster/issues/245) and tell me if it helps.

**Reply by txrubel** [reply]
Actually I didnt found any port conflict on my installation process. My running version is 0.2.20,

![image](https://user-images.githubusercontent.com/38957915/142948397-8c2f8775-5369-4c2f-b6ec-8ab879dd75ac.png)

**Reply by psanders** [reply]
I'm sorry you are having all these issues. This screenshot shows that the initialization process for the SIP Proxy failed. I believe that the initialization container ran before the SIP Proxy was ready.

At this point, all I can think is to [increase the preparation time](https://github.com/fonoster/fonoster/blob/1012c1c20b0f1e330ed7b009db928dc913517b8d/install.sh#L70). If it fixes the issue let me know, so we can find a better way to bootstrap the SIP Proxy.

**Reply by txrubel** [reply]
I have increased the wait timer to 60 sec, still same issue. I have changed some script on create_bucket_1 process, we just remove old bucket before creating new one. Is that create any issue?

I am guessing the issue could be on the auth process, I am not sure how the auth token initialized on sip_proxy. Are those initialized by the bootstrap process? If there any doc with them would be appreciable. 

As the providers container fired [BAD TOKEN] error, my wild guess is somehow this process getting wrong Token or proper token has not been initialized on the container.

The whole architecture is not clear to me, so please never mind about my wild guessing.

Screen shot after 60s wait,
![image](https://user-images.githubusercontent.com/38957915/143088217-2c943087-f367-4d1f-87e0-ac94d295c1c7.png)

**Reply by psanders** [reply]
> Are those initialized by the bootstrap process? If there any doc with them would be appreciable.

Exactly. The bootstrapping process ([the last thing we do](https://github.com/fonoster/fonoster/blob/1012c1c20b0f1e330ed7b009db928dc913517b8d/install.sh#L73)) creates the credentials on the SIP Server and also the storage buckets.

**Reply by txrubel** [reply]
so far I have checked in the bootstrap process first trying to login via this command,
` rctl login https://${SIPPROXY_HOST}:${SIPPROXY_API_PORT}/api/v1beta1 -u ${SIPPROXY_API_USERNAME} -p ${SIPPROXY_API_SECRET};`

then trying to run `rctl create -f /bootstrap.yml;`

According to the log, "Unable to retrieve token. Please verify your credentials" means the credentials not configured properly the running container and as this login process is failed the next create command also failed.

We tried to login via rest api it also not working and as we mentioned before [here](https://github.com/fonoster/fonoster/discussions/242#discussioncomment-1675276) the API_SECRET was wrongly configured.

Could you please guide me how we can debug the data_source in the sipproxy container and how the api secret is being configured in the data source?

**Reply by txrubel** [reply]
Here attached the sipproxy log, there have 2 exception, may be those are related one,
[fonoster_sipproxy_1.log](https://github.com/fonoster/fonoster/files/7591775/fonoster_sipproxy_1.log)

**Reply by txrubel** [reply]
here if we disabled the dynamic sipproxy api secret generation, then everything in docker loaded fine and even we can access sipproxy rest-api using that static api secret. But the issue still persist in providers create.

![image](https://user-images.githubusercontent.com/38957915/143185479-d17d9d50-de51-4480-ab1f-f7da6ee758ff.png)

Errors in providers container while we run, `fonoster providers:list`
![image](https://user-images.githubusercontent.com/38957915/143185687-a65bf6a0-97c6-403f-acfe-a66b7e9d4d8c.png)

For which key its showing this error? could you please help us to debug this issue?

**Reply by psanders** [reply]
It's a bit hard to help when I can reproduce the issue. My recommendation is for you start from scratch with a clean Docker Engine. Remove the folder fonoster created and if possible clean all the volumes. Then, install Fonoster and if it fails send all logs from the installation process.

> Here attached the sipproxy log, there have 2 exception, may be those are related one,

This should be a bug report, but I don't believe is the cause of the issue since the container is still responding.

**Reply by txrubel** [reply]
Dear @psanders , Cleanup docker help me to solve this issue. Thanks a lot for all the inputs.

**Reply by psanders** [reply]
I'm glad it worked! I marked this as answered, but feel free to comment if you have any related issues.
